jobsdb/migration.go

```diff
@@ -0,0 +1,479 @@
+			break
+		}
+
+		migrate, isSmall, recordsLeft := jd.checkIfMigrateDS(ds)
+		jd.logger.Debugf(
+			"[[ migrateDSLoop ]]: Migrate check %v, is small: %v, records left: %d, ds: %v",
+			migrate, isSmall, recordsLeft, ds,
+		)
+
+		if migrate {
+			if waiting != nil { // add current and waiting DS, no matter if the current ds is small or not, it doesn't matter
+				migrateFrom = append(migrateFrom, waiting.ds, ds)
+				insertBeforeDS = dsList[idx+1]
+				pendingJobsCount += waiting.recordsLeft + recordsLeft
+				liveDSCount += 2
+				waiting = nil
+			} else if !isSmall || len(migrateFrom) > 0 { // add only if the current DS is not small or if we already have some DS in the list
+				migrateFrom = append(migrateFrom, ds)
+				insertBeforeDS = dsList[idx+1]
+				pendingJobsCount += recordsLeft
+				liveDSCount++
+			} else { // add the current small DS as waiting for the next iteration to pickup
+				waiting = &smallDS{ds: ds, recordsLeft: recordsLeft}
+			}
+		} else {
+			waiting = nil // if there was a small DS waiting, we should remove it since its next dataset is not eligible for migration
+			if liveDSCount > 0 || migrateDSProbeCount > maxMigrateDSProbe {
+				// DS is not eligible for migration. But there are data sets on the left eligible to migrate, so break.
+				break
+			}
+		}
+		migrateDSProbeCount++
+	}
+	return
+}
+
+func (jd *HandleT) migrateJobsInTx(ctx context.Context, tx *Tx, srcDS, destDS dataSetT) (int, error) {
+	queryStat := stats.Default.NewTaggedStat("migration_jobs", stats.TimerType, stats.Tags{"customVal": jd.tablePrefix})
+	queryStat.Start()
+	defer queryStat.End()
+
+	compactDSQuery := fmt.Sprintf(
+		`with last_status as (select * from "v_last_%[1]s"),
+		inserted_jobs as
+		(
+			insert into %[3]q (job_id,   workspace_id,   uuid,   user_id,   custom_val,   parameters,   event_payload,   event_count,   created_at,   expire_at) 
+			           (select j.job_id, j.workspace_id, j.uuid, j.user_id, j.custom_val, j.parameters, j.event_payload, j.event_count, j.created_at, j.expire_at from %[2]q j left join last_status js on js.job_id = j.job_id
+				where js.job_id is null or js.job_state = ANY('{%[5]s}') order by j.job_id) returning job_id
+		),
+		insertedStatuses as 
+		(
+			insert into %[4]q (job_id, job_state, attempt, exec_time, retry_time, error_code, error_response, parameters) 
+			           (select job_id, job_state, attempt, exec_time, retry_time, error_code, error_response, parameters from last_status where job_state = ANY('{%[5]s}'))
+		)
+		select count(*) from inserted_jobs;`,
+		srcDS.JobStatusTable,
+		srcDS.JobTable,
+		destDS.JobTable,
+		destDS.JobStatusTable,
+		strings.Join(validNonTerminalStates, ","),
+	)
+
+	var numJobsMigrated int64
+	if err := tx.QueryRowContext(
+		ctx,
+		compactDSQuery,
+	).Scan(&numJobsMigrated); err != nil {
+		return 0, err
+	}
+	if _, err := tx.Exec(fmt.Sprintf(`ANALYZE %q, %q`, destDS.JobTable, destDS.JobStatusTable)); err != nil {
```

Comment:
If this returns error, there will be a problem.
1. We would end up with gw_1 & gw_1_1 tables
2. There will be an entry in journal table with done as FALSE. This will cause some other problems.

Should we ignore this error? @Sidddddarth @atzoum

Reply:
`WithTx` wrapper helps us avoid this situation.
All the suggested updates would've been rolled back.